consensus/core/src/tests/base_committer_tests.rs

```diff
@@ -0,0 +1,744 @@
+        assert_eq!(committed_block.author(), leader.authority)
+    } else {
+        panic!("Expected a committed leader")
+    };
+}
+
+/// Commit one by one each leader as the dag progresses in ideal conditions.
+#[test]
+#[tracing_test::traced_test]
+fn multiple_direct_commit() {
+    // Commitee of 4 with even stake
+    let context = Arc::new(Context::new_for_test(Some(4)));
+    let dag_state = Arc::new(RwLock::new(DagState::new(
+        context.clone(),
+        Arc::new(MemStore::new()),
+    )));
+    let committer = BaseCommitterBuilder::new(context.clone(), dag_state.clone()).build();
+
+    let wave_length = DEFAULT_WAVE_LENGTH;
+    let mut ancestors = None;
+
+    for n in 1..=10 {
+        let round_with_enough_blocks = wave_length * n;
+        ancestors = Some(build_dag(
+            context.clone(),
+            dag_state.clone(),
+            ancestors,
+            round_with_enough_blocks,
+        ));
+
+        // Leader round is the first round of each wave.
+        let leader_round = round_with_enough_blocks - wave_length;
+        let leader = committer
+            .elect_leader(leader_round)
+            .expect("should have elected leader");
+        tracing::info!("Try direct commit for leader {leader}",);
+        let leader_status = committer.try_direct_decide(leader);
+        tracing::info!("Leader commit status: {leader_status}");
+
+        if let LeaderStatus::Commit(ref committed_block) = leader_status {
+            assert_eq!(committed_block.author(), leader.authority)
+        } else {
+            panic!("Expected a committed leader")
+        };
+    }
+}
+
+/// We directly skip the leader if it has enough blame.
+#[test]
+#[tracing_test::traced_test]
+fn direct_skip() {
+    // Commitee of 4 with even stake
+    let context = Arc::new(Context::new_for_test(Some(4)));
+    let dag_state = Arc::new(RwLock::new(DagState::new(
+        context.clone(),
+        Arc::new(MemStore::new()),
+    )));
+    let committer = BaseCommitterBuilder::new(context.clone(), dag_state.clone()).build();
+
+    let wave_length = DEFAULT_WAVE_LENGTH;
+
+    // Add enough blocks to reach the leader of wave 1.
+    let voting_round_wave_1 = wave_length + 1;
+    let references_round_2 = build_dag(
+        context.clone(),
+        dag_state.clone(),
+        None,
+        voting_round_wave_1,
+    );
+
+    // Filter out that leader.
+    let leader_wave_1 = committer
+        .elect_leader(voting_round_wave_1 - 1)
+        .expect("should have elected leader");
+    let references_without_leader_wave_1: Vec<_> = references_round_2
+        .into_iter()
+        .filter(|x| x.author != leader_wave_1.authority)
+        .collect();
+
+    // Add enough blocks to reach the decision round of wave 1.
+    let leader_round_wave_2 = 2 * wave_length;
+    build_dag(
+        context.clone(),
+        dag_state.clone(),
+        Some(references_without_leader_wave_1),
+        leader_round_wave_2,
```

Comment:
I think - and it could be completely me - that having the `build_dag` produce blocks up to the `stop` exclusive can be a bit confusing when creating the tests. For instance here I believe it would be easier for me to understand something like:
```
build_dag(
        context.clone(),
        dag_state.clone(),
        Some(references_without_leader_wave_1),
        decision_round_wave_1,
```

whereas now I need to think across 2 waves, where in fact the second wave doesn't play any role here. In the end is probably a matter of preference, but I believe that this test could be written without any reference to  wave_2 and deal only with wave_1 which is the wave in question here.

Reply:
I am glad you brought this up. Kept going back and forth on what seemed to make more sense. I think because the rounds were zero indexed, it made some parts a little more confusing. But if you feel its easier to understand with stop being inclusive then I will change it back to that behavior.

Reply:
Yeah I think so - again to me - that with stop exclusive things might be a little easier to understand. Don't want this though to be a blocker for the PR.

Reply:
I refactored it 👍🏽